# 第一章：数据载入及初步观察

## 安装numpy和pandas包

In [1]:
#安装numpy和pandas
import os
os.system("pip install --upgrade pip")
os.system("pip install numpy==1.19.5")
os.system("pip install pandas==1.1.5")

0

In [2]:
import pandas as pd
import numpy as np
print(np.__version__)
print(pd.__version__)

1.19.5
1.1.5


如果显示的版本不是安装的版本，重启一下kernel就好了

In [3]:
#导入numpy和pandas
import numpy as np
import pandas as pd

## 载入数据

In [4]:
#使用相对路径载入数据
df = pd.read_csv('data/train.csv')
df.head(3)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S

In [5]:
#查看当前路径
import os
print(os.getcwd())

/home/jeffding/mycode/git/LearningCode/Datawhale/HandsOnDataAnalysis_2021June


In [6]:
#使用绝对路径载入数据
root_path="/home/jeffding/mycode/git/LearningCode/Datawhale/HandsOnDataAnalysis_2021June/"
df = pd.read_csv(root_path+'data/train.csv')
df.head(3)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S

**pd.read_csv()和pd.read_table()的不同**  
read_csv会自动识别分隔符进行分隔且只能读取csv文件，read_table需要添加一下分隔符，read_table可以读取非csv文件，只需要使用sep指定分隔符，就可以实现自动分隔了

In [7]:
#使用read_table载入数据
df = pd.read_table('data/train.csv')
df.head(3)

PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0  1,0,3,"Braund, Mr. Owen Harris",male,22,1,0,A/...                             
1  2,1,1,"Cumings, Mrs. John Bradley (Florence Br...                             
2  3,1,3,"Heikkinen, Miss. Laina",female,26,0,0,S...

In [8]:
#使用read_table载入数据（指定sep分隔符）
df = pd.read_table('data/train.csv',sep=',')
df.head(3)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S

In [9]:
#每1000行为一个数据模块，逐块读取
chunker = pd.read_csv('data/train.csv', chunksize=1000)

In [10]:
for i in chunker:
    print(i)

     PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ... 

**将表头改成中文，索引改为乘客ID [对于某些英文资料，我们可以通过翻译来更直观的熟悉我们的数据]**

```
PassengerId => 乘客ID  
Survived => 是否幸存  
Pclass => 乘客等级(1/2/3等舱位)  
Name => 乘客姓名  
Sex => 性别  
Age => 年龄  
SibSp => 堂兄弟/妹个数
Parch => 父母与小孩个数
Ticket => 船票信息
Fare => 票价
Cabin => 客舱
Embarked => 登船港口
```

In [11]:
df = pd.read_csv('data/train.csv', names=['乘客ID','是否幸存','仓位等级','姓名','性别','年龄','兄弟姐妹个数','父母子女个数','船票信息','票价','客舱','登船港口'],index_col='乘客ID',header=0)
df.head()

是否幸存  仓位等级                                                 姓名      性别  \
乘客ID                                                                          
1        0     3                            Braund, Mr. Owen Harris    male   
2        1     1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female   
3        1     3                             Heikkinen, Miss. Laina  female   
4        1     1       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female   
5        0     3                           Allen, Mr. William Henry    male   

        年龄  兄弟姐妹个数  父母子女个数              船票信息       票价    客舱 登船港口  
乘客ID                                                              
1     22.0       1       0         A/5 21171   7.2500   NaN    S  
2     38.0       1       0          PC 17599  71.2833   C85    C  
3     26.0       0       0  STON/O2. 3101282   7.9250   NaN    S  
4     35.0       1       0            113803  53.1000  C123    S  
5     35.0       0       0            373450   8.0500   NaN    S

## 初步观察

In [12]:
#查看数据的基本信息
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 891 entries, 1 to 891
Data columns (total 11 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   是否幸存    891 non-null    int64  
 1   仓位等级    891 non-null    int64  
 2   姓名      891 non-null    object 
 3   性别      891 non-null    object 
 4   年龄      714 non-null    float64
 5   兄弟姐妹个数  891 non-null    int64  
 6   父母子女个数  891 non-null    int64  
 7   船票信息    891 non-null    object 
 8   票价      891 non-null    float64
 9   客舱      204 non-null    object 
 10  登船港口    889 non-null    object 
dtypes: float64(2), int64(4), object(5)
memory usage: 83.5+ KB


In [13]:
#观察表格前10行的数据
df.head(10)

是否幸存  仓位等级                                                 姓名      性别  \
乘客ID                                                                          
1        0     3                            Braund, Mr. Owen Harris    male   
2        1     1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female   
3        1     3                             Heikkinen, Miss. Laina  female   
4        1     1       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female   
5        0     3                           Allen, Mr. William Henry    male   
6        0     3                                   Moran, Mr. James    male   
7        0     1                            McCarthy, Mr. Timothy J    male   
8        0     3                     Palsson, Master. Gosta Leonard    male   
9        1     3  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female   
10       1     2                Nasser, Mrs. Nicholas (Adele Achem)  female   

        年龄  兄弟姐妹个数  父母子女个数              船票信息       票价    客舱 登船港口  
乘客ID                                                              
1     22.0       1       0         A/5 21171   7.2500   NaN    S  
2     38.0       1       0          PC 17599  71.2833   C85    C  
3     26.0       0       0  STON/O2. 3101282   7.9250   NaN    S  
4     35.0       1       0            113803  53.1000  C123    S  
5     35.0       0       0            373450   8.0500   NaN    S  
6      NaN       0       0            330877   8.4583   NaN    Q  
7     54.0       0       0             17463  51.8625   E46    S  
8      2.0       3       1            349909  21.0750   NaN    S  
9     27.0       0       2            347742  11.1333   NaN    S  
10    14.0       1       0            237736  30.0708   NaN    C

In [14]:
#观察表格后15行的数据
df.tail(15)

是否幸存  仓位等级                                             姓名      性别    年龄  \
乘客ID                                                                            
877      0     3                  Gustafsson, Mr. Alfred Ossian    male  20.0   
878      0     3                           Petroff, Mr. Nedelio    male  19.0   
879      0     3                             Laleff, Mr. Kristo    male   NaN   
880      1     1  Potter, Mrs. Thomas Jr (Lily Alexenia Wilson)  female  56.0   
881      1     2   Shelley, Mrs. William (Imanita Parrish Hall)  female  25.0   
882      0     3                             Markun, Mr. Johann    male  33.0   
883      0     3                   Dahlberg, Miss. Gerda Ulrika  female  22.0   
884      0     2                  Banfield, Mr. Frederick James    male  28.0   
885      0     3                         Sutehall, Mr. Henry Jr    male  25.0   
886      0     3           Rice, Mrs. William (Margaret Norton)  female  39.0   
887      0     2                          Montvila, Rev. Juozas    male  27.0   
888      1     1                   Graham, Miss. Margaret Edith  female  19.0   
889      0     3       Johnston, Miss. Catherine Helen "Carrie"  female   NaN   
890      1     1                          Behr, Mr. Karl Howell    male  26.0   
891      0     3                            Dooley, Mr. Patrick    male  32.0   

      兄弟姐妹个数  父母子女个数              船票信息       票价    客舱 登船港口  
乘客ID                                                        
877        0       0              7534   9.8458   NaN    S  
878        0       0            349212   7.8958   NaN    S  
879        0       0            349217   7.8958   NaN    S  
880        0       1             11767  83.1583   C50    C  
881        0       1            230433  26.0000   NaN    S  
882        0       0            349257   7.8958   NaN    S  
883        0       0              7552  10.5167   NaN    S  
884        0       0  C.A./SOTON 34068  10.5000   NaN    S  
885        0       0   SOTON/OQ 392076   7.0500   NaN    S  
886        0       5            382652  29.1250   NaN    Q  
887        0       0            211536  13.0000   NaN    S  
888        0       0            112053  30.0000   B42    S  
889        1       2        W./C. 6607  23.4500   NaN    S  
890        0       0            111369  30.0000  C148    C  
891        0       0            370376   7.7500   NaN    Q

In [15]:
#判断数据是否为空，为空的地方返回True，其余地方返回False
df.isnull().head()

是否幸存   仓位等级     姓名     性别     年龄  兄弟姐妹个数  父母子女个数   船票信息     票价     客舱  \
乘客ID                                                                           
1     False  False  False  False  False   False   False  False  False   True   
2     False  False  False  False  False   False   False  False  False  False   
3     False  False  False  False  False   False   False  False  False   True   
4     False  False  False  False  False   False   False  False  False  False   
5     False  False  False  False  False   False   False  False  False   True   

       登船港口  
乘客ID         
1     False  
2     False  
3     False  
4     False  
5     False

In [16]:
import os
if not os.path.exists('result'):
    os.makedirs('./result')
df.to_csv('result/train_chinese.csv')

## Pandas 基础

pandas中有两个数据类型DateFrame和Series，通过查找简单了解他们。然后自己写一个关于这两个数据类型的小例子

In [17]:
#使用Series
import numpy as np
import pandas as pd
sdata = {'Ohio': 35000, 'Texas': 71000, 'Oregon': 16000, 'Utah': 5000}
example_1 = pd.Series(sdata)
example_1

Ohio      35000
Texas     71000
Oregon    16000
Utah       5000
dtype: int64

In [18]:
#使用DataFrame
data = {'state': ['Ohio', 'Ohio', 'Ohio', 'Nevada', 'Nevada', 'Nevada'],
        'year': [2000, 2001, 2002, 2001, 2002, 2003],'pop': [1.5, 1.7, 3.6, 2.4, 2.9, 3.2]}
example_2 = pd.DataFrame(data)
example_2

state  year  pop
0    Ohio  2000  1.5
1    Ohio  2001  1.7
2    Ohio  2002  3.6
3  Nevada  2001  2.4
4  Nevada  2002  2.9
5  Nevada  2003  3.2

In [19]:
#载入数据
df = pd.read_csv('data/train.csv')
df.head(3)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S

In [20]:
#查看表格所有的列名
df.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

In [21]:
#查看"cabin"这列的所有项
#方法一
df['Cabin'].head(3)

0    NaN
1    C85
2    NaN
Name: Cabin, dtype: object

In [22]:
#方法二
df.Cabin.head(3)

0    NaN
1    C85
2    NaN
Name: Cabin, dtype: object

In [23]:
#加载文件"test_1.csv"，然后对比"train.csv"，看看有哪些多出的列，然后将多出的列删除
test_1 = pd.read_csv('data/test_1.csv')
test_1.head(3)

Unnamed: 0  PassengerId  Survived  Pclass  \
0           0            1         0       3   
1           1            2         1       1   
2           2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked    a  
0      0         A/5 21171   7.2500   NaN        S  100  
1      0          PC 17599  71.2833   C85        C  100  
2      0  STON/O2. 3101282   7.9250   NaN        S  100

In [24]:
# 使用del删除多余的列
del test_1['a']
test_1.head(3)

Unnamed: 0  PassengerId  Survived  Pclass  \
0           0            1         0       3   
1           1            2         1       1   
2           2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S

In [25]:
# 使用drop()删除多余的列
test_1 = pd.read_csv('data/test_1.csv')
test_1.drop('a',axis=1)

Unnamed: 0  PassengerId  Survived  Pclass  \
0             0            1         0       3   
1             1            2         1       1   
2             2            3         1       3   
3             3            4         1       1   
4             4            5         0       3   
..          ...          ...       ...     ...   
886         886          887         0       2   
887         887          888         1       1   
888         888          889         0       3   
889         889          890         1       1   
890         890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 13 columns]

In [26]:
#将['PassengerId','Name','Age','Ticket']这几个列元素隐藏，只观察其他几个列元素
df.drop(['PassengerId','Name','Age','Ticket'],axis=1).head(3)

Survived  Pclass     Sex  SibSp  Parch     Fare Cabin Embarked
0         0       3    male      1      0   7.2500   NaN        S
1         1       1  female      1      0  71.2833   C85        C
2         1       3  female      0      0   7.9250   NaN        S

In [27]:
df2 = pd.read_csv('data/train.csv')
df2.drop(['PassengerId','Name','Age','Ticket'],axis=1,inplace=True)

In [28]:
df2.head(3)

Survived  Pclass     Sex  SibSp  Parch     Fare Cabin Embarked
0         0       3    male      1      0   7.2500   NaN        S
1         1       1  female      1      0  71.2833   C85        C
2         1       3  female      0      0   7.9250   NaN        S

In [29]:
#以"Age"为筛选条件，显示年龄在10岁以下的乘客信息
df[df["Age"]<10].head(3)

PassengerId  Survived  Pclass                             Name     Sex  \
7             8         0       3   Palsson, Master. Gosta Leonard    male   
10           11         1       3  Sandstrom, Miss. Marguerite Rut  female   
16           17         0       3             Rice, Master. Eugene    male   

    Age  SibSp  Parch   Ticket    Fare Cabin Embarked  
7   2.0      3      1   349909  21.075   NaN        S  
10  4.0      1      1  PP 9549  16.700    G6        S  
16  2.0      4      1   382652  29.125   NaN        Q

In [30]:
#以"Age"为条件，将年龄在10岁以上和50岁以下的乘客信息显示出来，并将这个数据命名为midage
midage = df[(df["Age"]>10)& (df["Age"]<50)]
midage.head(3)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S

**数据清洗时，会将带空值的行删除，此时DataFrame或Series类型的数据不再是连续的索引，可以使用reset_index()重置索引。**

In [31]:
#将midage的数据中第100行的"Pclass"和"Sex"的数据显示出来
midage = midage.reset_index(drop=True)
midage.head(3)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S

In [32]:
midage.loc[[100],['Pclass','Sex']]

Pclass   Sex
100       2  male

In [33]:
#使用loc方法将midage的数据中第100，105，108行的"Pclass"，"Name"和"Sex"的数据显示出来
midage.loc[[100,105,108],['Pclass','Name','Sex']] #因为你主动的延长了行的距离，所以会产生表格形式

Pclass                               Name   Sex
100       2  Byles, Rev. Thomas Roussel Davids  male
105       3           Cribb, Mr. John Hatfield  male
108       3                    Calic, Mr. Jovo  male

In [34]:
#使用iloc方法将midage的数据中第100，105，108行的"Pclass"，"Name"和"Sex"的数据显示出来
midage.iloc[[100,105,108],[2,3,4]]

Pclass                               Name   Sex
100       2  Byles, Rev. Thomas Roussel Davids  male
105       3           Cribb, Mr. John Hatfield  male
108       3                    Calic, Mr. Jovo  male

## 探索性数据分析

In [35]:
#载入之前保存的train_chinese.csv数据，关于泰坦尼克号的任务，我们就使用这个数据
text = pd.read_csv('result/train_chinese.csv')
text.head()

乘客ID  是否幸存  仓位等级                                                 姓名  \
0     1     0     3                            Braund, Mr. Owen Harris   
1     2     1     1  Cumings, Mrs. John Bradley (Florence Briggs Th...   
2     3     1     3                             Heikkinen, Miss. Laina   
3     4     1     1       Futrelle, Mrs. Jacques Heath (Lily May Peel)   
4     5     0     3                           Allen, Mr. William Henry   

       性别    年龄  兄弟姐妹个数  父母子女个数              船票信息       票价    客舱 登船港口  
0    male  22.0       1       0         A/5 21171   7.2500   NaN    S  
1  female  38.0       1       0          PC 17599  71.2833   C85    C  
2  female  26.0       0       0  STON/O2. 3101282   7.9250   NaN    S  
3  female  35.0       1       0            113803  53.1000  C123    S  
4    male  35.0       0       0            373450   8.0500   NaN    S

In [36]:
#利用Pandas对示例数据进行排序，要求升序
#自己构建一个都为数字的DataFrame数据
frame = pd.DataFrame(np.arange(8).reshape((2, 4)), 
                     index=['2', '1'], 
                     columns=['d', 'a', 'b', 'c'])
frame

d  a  b  c
2  0  1  2  3
1  4  5  6  7

【代码解析】

pd.DataFrame() ：创建一个DataFrame对象

np.arange(8).reshape((2, 4)) : 生成一个二维数组（2*4）,第一列：0，1，2，3 第二列：4，5，6，7

index=['2, 1] ：DataFrame 对象的索引列

columns=['d', 'a', 'b', 'c'] ：DataFrame 对象的索引行

In [37]:
# 大多数时候我们都是想根据列的值来排序,所以，将你构建的DataFrame中的数据根据某一列，升序排列
frame.sort_values(by='c', ascending=True) #ascending=True升序 False降序

d  a  b  c
2  0  1  2  3
1  4  5  6  7

In [38]:
frame.sort_values(by='c', ascending=False) #ascending=True升序 False降序

d  a  b  c
1  4  5  6  7
2  0  1  2  3

In [39]:
# 让行索引升序排序
frame.sort_index()

d  a  b  c
1  4  5  6  7
2  0  1  2  3

In [40]:
# 让列索引升序排序
frame.sort_index(axis=1)

a  b  c  d
2  1  2  3  0
1  5  6  7  4

In [41]:
# 让任选两列数据同时降序排序
frame.sort_values(by=['a', 'c'], ascending=False)

d  a  b  c
1  4  5  6  7
2  0  1  2  3

In [42]:
#对泰坦尼克号数据（trian.csv）按票价和年龄两列进行综合排序（降序排列），从数据中你能发现什么
text.sort_values(by=['票价', '年龄'], ascending=False).head(3)

乘客ID  是否幸存  仓位等级                                  姓名      性别    年龄  \
679   680     1     1  Cardeza, Mr. Thomas Drake Martinez    male  36.0   
258   259     1     1                    Ward, Miss. Anna  female  35.0   
737   738     1     1              Lesurer, Mr. Gustave J    male  35.0   

     兄弟姐妹个数  父母子女个数      船票信息        票价           客舱 登船港口  
679       0       1  PC 17755  512.3292  B51 B53 B55    C  
258       0       0  PC 17755  512.3292          NaN    C  
737       0       0  PC 17755  512.3292         B101    C

In [43]:
#利用Pandas进行算术计算，计算两个DataFrame数据相加结果
#建立一个例子
frame1_a = pd.DataFrame(np.arange(9.).reshape(3, 3),
                     columns=['a', 'b', 'c'],
                     index=['one', 'two', 'three'])
frame1_b = pd.DataFrame(np.arange(12.).reshape(4, 3),
                     columns=['a', 'e', 'c'],
                     index=['first', 'one', 'two', 'second'])
frame1_a

a    b    c
one    0.0  1.0  2.0
two    3.0  4.0  5.0
three  6.0  7.0  8.0

In [44]:
frame1_b

a     e     c
first   0.0   1.0   2.0
one     3.0   4.0   5.0
two     6.0   7.0   8.0
second  9.0  10.0  11.0

In [45]:
#将frame_a和frame_b进行相加
frame1_a + frame1_b

a   b     c   e
first   NaN NaN   NaN NaN
one     3.0 NaN   7.0 NaN
second  NaN NaN   NaN NaN
three   NaN NaN   NaN NaN
two     9.0 NaN  13.0 NaN

In [46]:
#通过泰坦尼克号数据如何计算出在船上最大的家族有多少人
'''
还是用之前导入的chinese_train.csv如果我们想看看在船上，最大的家族有多少人（‘兄弟姐妹个数’+‘父母子女个数’），我们该怎么做呢？
'''
max(text['兄弟姐妹个数'] + text['父母子女个数'])

10

In [47]:
#学会使用Pandas describe()函数查看数据基本统计信息
#建立一个例子
frame2 = pd.DataFrame([[1.4, np.nan], 
                       [7.1, -4.5],
                       [np.nan, np.nan], 
                       [0.75, -1.3]
                      ], index=['a', 'b', 'c', 'd'], columns=['one', 'two'])
frame2

one  two
a  1.40  NaN
b  7.10 -4.5
c   NaN  NaN
d  0.75 -1.3

In [48]:
# 调用 describe 函数，观察frame2的数据基本信息

frame2.describe()

one       two
count  3.000000  2.000000
mean   3.083333 -2.900000
std    3.493685  2.262742
min    0.750000 -4.500000
25%    1.075000 -3.700000
50%    1.400000 -2.900000
75%    4.250000 -2.100000
max    7.100000 -1.300000

```
count : 样本数据大小
mean : 样本数据的平均值
std : 样本数据的标准差
min : 样本数据的最小值
25% : 样本数据25%的时候的值
50% : 样本数据50%的时候的值
75% : 样本数据75%的时候的值
max : 样本数据的最大值
```

In [49]:
#分别看看泰坦尼克号数据集中 票价、父母子女 这列数据的基本统计数据，你能发现什么？
'''
看看泰坦尼克号数据集中 票价 这列数据的基本统计数据
'''
text['票价'].describe()

count    891.000000
mean      32.204208
std       49.693429
min        0.000000
25%        7.910400
50%       14.454200
75%       31.000000
max      512.329200
Name: 票价, dtype: float64

从上面数据我们可以看出， 一共有891个票价数据， 平均值约为：32.20， 标准差约为49.69，说明票价波动特别大， 25%的人的票价是低于7.91的，50%的人的票价低于14.45，75%的人的票价低于31.00， 票价最大值约为512.33，最小值为0。

In [50]:
'''
通过上面的例子，我们再看看泰坦尼克号数据集中 父母子女个数 这列数据的基本统计数据，然后可以说出你的想法
'''
text['父母子女个数'].describe()

count    891.000000
mean       0.381594
std        0.806057
min        0.000000
25%        0.000000
50%        0.000000
75%        0.000000
max        6.000000
Name: 父母子女个数, dtype: float64